<a href="https://colab.research.google.com/github/christianasb/Python-Capstone-DataCamp-Mini-Projects/blob/main/airbnb-market-trends/AirBnB_Market_Trends_in_NYC_Mini_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![NYC Skyline](img/nyc.jpg)

Welcome to New York City, one of the most-visited cities in the world. There are many Airbnb listings in New York City to meet the high demand for temporary lodging for travelers, which can be anywhere between a few nights to many months. In this project, you will take a closer look at the New York Airbnb market by combining data from multiple file types like `.csv`, `.tsv`, and `.xlsx` (Excel files).

Recall that **CSV**, **TSV**, and **Excel** files are three common formats for storing data.
Three files containing data on 2019 Airbnb listings are available to you:

**data/airbnb_price.csv**
This is a CSV file containing data on Airbnb listing prices and locations.
- **`listing_id`**: unique identifier of listing
- **`price`**: nightly listing price in USD
- **`nbhood_full`**: name of borough and neighborhood where listing is located

**data/airbnb_room_type.xlsx**
This is an Excel file containing data on Airbnb listing descriptions and room types.
- **`listing_id`**: unique identifier of listing
- **`description`**: listing description
- **`room_type`**: Airbnb has three types of rooms: shared rooms, private rooms, and entire homes/apartments

**data/airbnb_last_review.tsv**
This is a TSV file containing data on Airbnb host names and review dates.
- **`listing_id`**: unique identifier of listing
- **`host_name`**: name of listing host
- **`last_review`**: date when the listing was last reviewed


## Loading & Examining Data

In [45]:
# Import required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import requests
from io import StringIO
from io import BytesIO

In [46]:
#Load CSV file

#read in the CSV data from a Google Drive shareable link
csv_orig_url="https://drive.google.com/file/d/1J2isp3aAIBSyGkEgAif3MuomfxXwlzdm/view?usp=sharing"
csv_file_id = csv_orig_url.split('/')[-2] #extract file ID from url
csv_dwn_url = 'https://drive.google.com/uc?export=download&id=' + csv_file_id

#download and convert CSV content into a readable format
csv_url = requests.get(csv_dwn_url).text
csv_raw = StringIO(csv_url)

#read the data into a pandas DataFrame
df_price = pd.read_csv(csv_raw)

#examine first 5 rows of CSV price data
df_price.head()

listing_id        price                nbhood_full
0        2595  225 dollars         Manhattan, Midtown
1        3831   89 dollars     Brooklyn, Clinton Hill
2        5099  200 dollars     Manhattan, Murray Hill
3        5178   79 dollars  Manhattan, Hell's Kitchen
4        5238  150 dollars       Manhattan, Chinatown

In [47]:
#check price columns and data types
df_price.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25209 entries, 0 to 25208
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   listing_id   25209 non-null  int64 
 1   price        25209 non-null  object
 2   nbhood_full  25209 non-null  object
dtypes: int64(1), object(2)
memory usage: 591.0+ KB


In [48]:
#summary statistics for price numeric columns
df_price.describe()

listing_id
count  2.520900e+04
mean   2.068922e+07
std    1.102928e+07
min    2.595000e+03
25%    1.202273e+07
50%    2.234391e+07
75%    3.037669e+07
max    3.645581e+07

In [49]:
#Load Excel file

#read in the Excel data from a Google Drive shareable link
excel_orig_url="https://docs.google.com/spreadsheets/d/1sqlyXgd1J6CjHlIsd4oYHeLNjnBcJreE/edit?usp=sharing&ouid=113904096852535515268&rtpof=true&sd=true"
excel_file_id = excel_orig_url.split('/')[-2] #extract file ID from url
excel_dwn_url = 'https://drive.google.com/uc?export=download&id=' + excel_file_id

#download the file content as bytes
excel_url = requests.get(excel_dwn_url)
excel_raw = BytesIO(excel_url.content)

#read the data into a pandas DataFrame
df_room_type = pd.read_excel(excel_raw)

#examine first 5 rows of Excel room_type data
df_room_type.head()

listing_id                                description        room_type
0        2595                      Skylit Midtown Castle  Entire home/apt
1        3831            Cozy Entire Floor of Brownstone  Entire home/apt
2        5099  Large Cozy 1 BR Apartment In Midtown East  Entire home/apt
3        5178            Large Furnished Room Near B'way     private room
4        5238         Cute & Cozy Lower East Side 1 bdrm  Entire home/apt

In [50]:
#check room_type columns and data types
df_room_type.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25209 entries, 0 to 25208
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   listing_id   25209 non-null  int64 
 1   description  25199 non-null  object
 2   room_type    25209 non-null  object
dtypes: int64(1), object(2)
memory usage: 591.0+ KB


In [51]:
#examine number of unique room_type values
print(df_room_type['room_type'].unique())

['Entire home/apt' 'private room' 'Private room' 'entire home/apt'
 'PRIVATE ROOM' 'shared room' 'ENTIRE HOME/APT' 'Shared room'
 'SHARED ROOM']


In [52]:
#Load TSV file

#read in the TSV data from a Google Drive shareable link
tsv_orig_url="https://drive.google.com/file/d/1jrjGrzVeqeFFlgZ7HVoWKgRlmGtm3Um7/view?usp=sharing"
tsv_file_id = tsv_orig_url.split('/')[-2] #extract file ID from url
tsv_dwn_url = 'https://drive.google.com/uc?export=download&id=' + tsv_file_id

#download and convert TSV content into a readable format
tsv_url = requests.get(tsv_dwn_url).text
tsv_raw = StringIO(tsv_url)

#read the data into a pandas DataFrame
df_last_review = pd.read_csv(tsv_raw, sep='\t')

#examine first 5 rows of TSV last_review data
df_last_review.head()

listing_id    host_name   last_review
0        2595     Jennifer   May 21 2019
1        3831  LisaRoxanne  July 05 2019
2        5099        Chris  June 22 2019
3        5178     Shunichi  June 24 2019
4        5238          Ben  June 09 2019

In [53]:
#check last_review column and data types
df_last_review.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25209 entries, 0 to 25208
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   listing_id   25209 non-null  int64 
 1   host_name    25201 non-null  object
 2   last_review  25209 non-null  object
dtypes: int64(1), object(2)
memory usage: 591.0+ KB


## Merging the 3 Different DataFrames

In [54]:
#since each DataFrame shares the column "listing id," I will use this to merge them

#first merge df_price and df_room_type on "listing_id"
merged_df = pd.merge(df_price, df_room_type, on='listing_id')

#then merge the result with df_last_review on "listing_id"
merged_df = pd.merge(merged_df, df_last_review, on='listing_id')

#check the merged DataFrame
print(merged_df.head())

   listing_id        price                nbhood_full  \
0        2595  225 dollars         Manhattan, Midtown   
1        3831   89 dollars     Brooklyn, Clinton Hill   
2        5099  200 dollars     Manhattan, Murray Hill   
3        5178   79 dollars  Manhattan, Hell's Kitchen   
4        5238  150 dollars       Manhattan, Chinatown   

                                 description        room_type    host_name  \
0                      Skylit Midtown Castle  Entire home/apt     Jennifer   
1            Cozy Entire Floor of Brownstone  Entire home/apt  LisaRoxanne   
2  Large Cozy 1 BR Apartment In Midtown East  Entire home/apt        Chris   
3            Large Furnished Room Near B'way     private room     Shunichi   
4         Cute & Cozy Lower East Side 1 bdrm  Entire home/apt          Ben   

    last_review  
0   May 21 2019  
1  July 05 2019  
2  June 22 2019  
3  June 24 2019  
4  June 09 2019  


## Determining the Earliest and Most Recent Review Dates

In [55]:
#convert relevant reviews column to date format
merged_df['last_review'] = pd.to_datetime(merged_df['last_review'], errors='coerce')

#find the earliest review date
first_reviewed = merged_df['last_review'].min()

#find the latest review date
last_reviewed = merged_df['last_review'].max()

#print the results
print(f"Earliest review date: {first_reviewed}")
print(f"Most recent review date: {last_reviewed}")

Earliest review date: 2019-01-01 00:00:00
Most recent review date: 2019-07-09 00:00:00


## Finding How Many Listings are Private Rooms

In [56]:
#convert all room_type values to lowercase to standardize them
merged_df['room_type'] = merged_df['room_type'].str.lower()

#filter the DataFrame for rows where room_type is "private room"
private_rooms_df = merged_df[merged_df['room_type'] == 'private room']

#count how many listings are private rooms
nb_private_rooms = private_rooms_df.shape[0]

#print the result
print(f"Number of private room listings: {nb_private_rooms}")

Number of private room listings: 11356


## Finding the Average Price of Listings

In [57]:
#remove text/symbols like "dollars" or "$" using str.replace to standardize prices
merged_df['price_cleaned'] = merged_df['price'].str.replace('[^\d.]', '', regex=True) #The regex pattern '[^\d.]' removes any character that's not a digit or dot

#convert the cleaned string values to float
merged_df['price_cleaned'] = merged_df['price_cleaned'].astype(float)

#calculate the average price and round to 2 decimal places
avg_price = round(merged_df['price_cleaned'].mean(), 2)

#print the result
print(f"Average listing price: ${avg_price}")


Average listing price: $141.78


## Creating a DataFrame with the Four Solution Values

In [58]:
df_review_dates = pd.DataFrame({
    'first_reviewed': [first_reviewed],
    'last_reviewed': [last_reviewed],
    'nb_private_rooms': [nb_private_rooms],
    'avg_price': [avg_price]
})

# Display the result
print(df_review_dates)

  first_reviewed last_reviewed  nb_private_rooms  avg_price
0     2019-01-01    2019-07-09             11356     141.78


## Interpretation & Recommendations for NYC Airbnb Market Strategy

Based on the analysis of listing prices, room types, and review activity across New York City neighborhoods, I suggest the following strategic actions:


**Key Findings:**
- **Average listing price:** $141.78 per night, indicating a mid-range market with both budget and premium offerings.

- **High proportion of private rooms:** 11,356 out of 25,209 listings (~45%) are private rooms, reflecting strong demand for affordable, semi-private stays.

- **Review activity timeline:** Listings were actively reviewed between January 1, 2019, and July 9, 2019, highlighting a six-month snapshot of peak engagement.


**Recommended Actions:**
- **Capitalize on private room demand:** Promote and prioritize private room listings in search results and marketing campaigns to attract solo travelers, students, and budget tourists.

- **Support pricing strategy with data tools:** Equip hosts with tools to set competitive nightly rates based on borough, room type, and demand seasonality to maximize occupancy while maintaining value.

- **Engage inactive or low-review hosts:** Reconnect with hosts whose listings haven’t received reviews since early 2019. Offer incentives to update listing content and improve guest engagement.

- **Balance tourist traffic across neighborhoods:** Use borough-level insights to promote underbooked neighborhoods (e.g., parts of Queens, Bronx, Staten Island) and reduce pressure on high-density areas like Manhattan.